presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java

```diff
@@ -243,6 +243,11 @@ public boolean isCrossJoin()
         return criteria.isEmpty() && !filter.isPresent() && type == INNER;
     }
 
+    public boolean isSpatialJoin()
+    {
+        return criteria.isEmpty() && filter.isPresent() && type == INNER;
```

Comment:
This would return true for any non-spatial inequality inner join, no?

Reply:
@martint My understanding is that INNER join can only be a hash join or a cross join. To become hash join it must have at least one equi-clause. If there are no equi-clauses, it is a cross join. With this change, some cross joins become spatial joins. Hence, if an inner join has no equi-clauses, then it is either cross join or spatial join. If the join also has a filter, it must be a spatial join because cross join doesn't support filters.

Reply:
Please add relevant explanation to the source code too.

However, i think this is not fully correct. `com.facebook.presto.sql.planner.RelationPlanner#visitJoin` can create `JoinNode` with a filter. Then, `PredicatePushdown` looks responsible of getting things straight: https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java#L403
This means that reliability of `JoinNode.isSpatialJoin()` varies during planning :(

Reply:
Also, there should be some provision preventing the spatial join from de-spatial-izing, if `PredicatePushDown` is run after `TransformSpatialPredicateToJoin`. Even if this is not the case today, `PredicatePushDown` is a useful optimizer, which should be run more (iterative or more times).

I'd suggest at least considering adding an explicit boolean flag like `JoinNode#isSpatial`, or at least adding some control code to `PredicatePushDown` (nor sure how this should look like).

Reply:
?